mon: implement mon backup mechanism

poelzi opened the pull request:
Fixes https://tracker.ceph.com/issues/63801

This patchset implements ceph mon backup functionality using the rocksdb BackupEngine feature.

Manual backups can be triggered via the admin socket using the "backup" command.

If the new setting `mon.auto_backup` is enabled, the backup is triggered when:

- new OSD with encryption keys are added - OSD with encryption are note recoverable when mon is corrupted
- important auth permission changes - auth keys are not stored in OSDs


Backups can be restored with the `--restore-backup` and `--backup-version` arguments.
ceph-mon exits with 0 after successful restoration so the monitor can be started normally afterwards.
If no backup-version is specified, the last valid backup is restored.

Open questions: should the backup() called from the tick() function be run in the `ThreadPool cpu_tp` ?

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

poelzi committed:
mon: add read_meta_path

Add function to read simple meta data from
specified directroy without opening the store

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

frittentheke commented:
@poelzi i this really stale?

poelzi commented:
@frittentheke no, I just had a accident and will be back finishing this up in 2 weeks

frittentheke commented:
> @frittentheke no, I just had a accident and will be back finishing this up in 2 weeks

Oh, I hope you are ok.

Thanks for your response and for working on the backup feature. Really looking forward to have sth built in.

poelzi commented:
@frittentheke I finished the backup / cleanup code and added documentation.
I tested the backup with artificially blown up mon database and the backup performance is very good. I noticed no delays to mon quorum desync.

The backup cleanup mechanism implement last n backups, hourly and daily versions. It now also checks for free space before creating backups

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+==============
+
+The monitor daemon is capable of creating consistant backups of the RocksDB to allow restorating
+in case of fatal multi node failure. It uses the native RocksDB backup functionality to create consistant 
```
The Ceph Monitor can create consistent backups of its RocksDB to facilitate restoration in case the entire Monitor cluster fails.

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+When :ref:`Auto Backup <mon_auto_backup>` is enabled, creation of certain objects like auth
+permissions trigger a backup.
+
+To reduce IO on mon database, shared backups are the default, but every 5th backup or manually triggered backups are full backups.
```
In order to limit demands on the Ceph Monitor database, shared backups are the default.  Every fifth backup, or any manually-triggered backup however will be a full backup.

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+
+To reduce IO on mon database, shared backups are the default, but every 5th backup or manually triggered backups are full backups.
+If :ref:`Always Full Backup <mon_backup_always_full>` is enabled, all backups are full copies. Full backups can be copied to an external
+drive by copying the backup folder. Shared backus use a shared object tree and only safe pointers.
```
s/drive/location/
s/safe/save/?

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+The cleanup algorithm keeps the last ``mon_backup_keep_last`` backups. It then collects hourly ``mon_backup_keep_hourly``
+and daily ``mon_backup_keep_daily`` versions. It selects the largest of all variants in a time window.
+
+You can trigger `backup` and `backup_cleanup` through the admin socket
```
The socket of the lead mon? Or any mon?

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+  .. prompt:: bash #
+     ceph --admin-daemon .../mon.asok backup_cleanup
+
+Since all backup operations run in its own threads, monitoring should be done using performance metrics.
```
In their own threads

anthonyeleven commented on doc/rados/configuration/mon-config-ref.rst:
```diff
@@ -600,6 +600,88 @@ is far outweighed by the number of accidental pool (and thus data) deletions it
+     ceph --admin-daemon .../mon.asok backup_cleanup
+
+Since all backup operations run in its own threads, monitoring should be done using performance metrics.
+Use the prometheus interface to monitor for `backup_failed` or missing `backup_started` counters.
```
Use Prometheus to monitor

anthonyeleven commented on doc/rados/troubleshooting/troubleshooting-mon.rst:
```diff
@@ -529,6 +529,25 @@ or::
+.. code-block:: bash
+  ceph-mon -i [num] --list-backups /path/to/backups
+
+To restore a backup version, run following command:
```
s/version//

anthonyeleven commented on doc/rados/troubleshooting/troubleshooting-mon.rst:
```diff
@@ -529,6 +529,25 @@ or::
+.. code-block:: bash
+  ceph-mon -i [num] --restore-backup /path/to/backups --backup-version <version>
+
+You can skip the `--backup-version` argument, which will use the latest version automatically.
```
If the `--backup-version` argument is omitted, the latest version will be restored.

anthonyeleven commented on doc/rados/troubleshooting/troubleshooting-mon.rst:
```diff
@@ -529,6 +529,25 @@ or::
+  ceph-mon -i [num] --restore-backup /path/to/backups --backup-version <version>
+
+You can skip the `--backup-version` argument, which will use the latest version automatically.
+The cluster will recover from that point on by applying missing `osdmap` versions stored on the osds.
```
versions to each OSD.

anthonyeleven commented on doc/rados/troubleshooting/troubleshooting-mon.rst:
```diff
@@ -529,6 +529,25 @@ or::
+
+You can skip the `--backup-version` argument, which will use the latest version automatically.
+The cluster will recover from that point on by applying missing `osdmap` versions stored on the osds.
+Auth accounts and other none osd related entries are lost.
```
other non-OSD-related

anthonyeleven commented on src/common/options/mon.yaml.in:
```diff
@@ -1091,6 +1091,90 @@ options:
+- name: mon_auto_backup
+  type: bool
+  level: advanced
+  desc: automatically trigger backups on important database changes
```
desc fields should be all Capitalized or all lowercase

athanatos commented on src/mon/Monitor.cc:
```diff
@@ -275,6 +276,8 @@ Monitor::Monitor(CephContext* cct_, string nm, MonitorDBStore *s,
   // retry (and revalidate).
   leader_mon_commands = local_mon_commands;
+  
+  backup_manager = new MonitorBackupManager(cct, this);
```
initializer list

athanatos commented on src/mon/Monitor.cc:
```diff
@@ -496,6 +504,50 @@ int Monitor::do_admin_command(
 
+int Monitor::backup(bool full)
+{
+  if(!backup_manager) {
```
Is it possible for this to be null?  space between if and '(' (and please check every other condition in this PR).

poelzi replied:
Test environments or admin socket while startup. I try to be rather save then sorry

athanatos commented on src/mon/Monitor.cc:
```diff
@@ -496,6 +504,50 @@ int Monitor::do_admin_command(
+
+int Monitor::backup_cleanup()
+{
+  if(!backup_manager) {
```
if (

athanatos commented on src/mon/Monitor.h:
```diff
@@ -1024,6 +1044,9 @@ class Monitor : public Dispatcher,
   OpTracker op_tracker;
 
+  MonitorBackupManager *backup_manager;
+  bool mon_backup_requested;
```
Initialize

athanatos commented on src/mon/Monitor.h:
```diff
@@ -1069,6 +1092,14 @@ class Monitor : public Dispatcher,
 		       std::ostream& out);
 
+  // Notify monitor that it should create a new database backup
+  void should_backup() {
```
set_should_backup

athanatos commented on src/mon/MonitorDBStore.h:
```diff
@@ -736,6 +743,61 @@ class MonitorDBStore
+  /// @brief Creates a backup of the database
+  /// @param backup_path location to create backup at
+  /// @return true on success
+  KeyValueDB::BackupStats backup(bool full = false) {
```
Does this actually need a default value?

poelzi replied:
When full is not set, incremental backup is likely created.

athanatos commented on src/mon/Monitor.h:
```diff
@@ -1024,6 +1044,9 @@ class Monitor : public Dispatcher,
 
   OpTracker op_tracker;
 
+  MonitorBackupManager *backup_manager;
```
Why is this a pointer and not just a directly contained value?  If you do actually need an indirection, please use a std::unique_ptr.

athanatos reviewed (changes requested):
Initial comments, still needs a more detailed review.

athanatos commented:
There also appear to be CephaCheck / ceph-pr-docs comments that need to be addressed?

I do like this feature -- regular/easy monitor backups seems like a solid addition.

athanatos commented:
Ping me once my above comments have been addressed and I'll find time to do a more detailed review.

poelzi commented:
@athanatos I fixed all review remarks

frittentheke commented:
@athanatos please dont let this one stale out .... would be awesome to have built in mon backups

ceph-jenkins commented:
Can one of the admins verify this patch?

poelzi commented:
Sorry, I don't work for b1-systems anymore. The patch is basically finished.
While testing, I experienced one corruption in the backup. It was an extra byte somewhere and I suspect, it is related to the  backup manager.

frittentheke commented:
@athanatos please don't let this awesome feature go to waste.

frittentheke commented:
@athanatos could you kindly take another peek so this PR and awesome feature does not go to waste?

pejotes commented:
yes, this is a great feature and highly appreciated from the community. Mon's are heart of our operations and we are always worried about them, having this last resort of recovery (even if having to recover from an osd is still there).

frittentheke commented:
Can we please dont let this go to waste?

tchaikov commented on src/kv/RocksDBStore.cc:
```diff
@@ -2014,7 +2014,15 @@ bool RocksDBStore::backup(const std::string& path)
     rocksdb::Env::Default(),
     &backup_engine);
+  if (!backup_engine || !s.ok()) {
+    derr << __func__ << "Can't create backup_engine: " << s.ToString() << dendl;
```
please squash the fix commits into the commit fixed by them, so the commit history is consistent.

tchaikov commented on src/mon/Monitor.h:
```diff
@@ -1029,6 +1050,9 @@ class Monitor : public Dispatcher,
 
   OpTracker op_tracker;
 
+  MonitorBackupManager *backup_manager = nullptr;
```
please use smart pointer in favor of raw pointer.

tchaikov commented on src/kv/RocksDBStore.cc:
```diff
@@ -2006,6 +2011,274 @@ int RocksDBStore::split_key(rocksdb::Slice in, string *prefix, string *key)
+
+bool compare_backupinfo_by_timestamp (rocksdb::BackupInfo a, rocksdb::BackupInfo b)
+{
+  return (a.timestamp >= b.timestamp);
```
please drop extraneous parenthesis around the statement.

tchaikov commented on src/kv/RocksDBStore.cc:
```diff
@@ -2006,6 +2011,274 @@ int RocksDBStore::split_key(rocksdb::Slice in, string *prefix, string *key)
+  utime_t start;
+  utime_t end;
+  rocksdb::BackupID backup_id;
+  uint64_t backup_size;
```
```suggestion
  rocksdb::BackupID backup_id = 0;
  uint64_t backup_size = 0;
```

tchaikov commented on src/kv/RocksDBStore.cc:
```diff
@@ -2006,6 +2011,274 @@ int RocksDBStore::split_key(rocksdb::Slice in, string *prefix, string *key)
+  std::vector<rocksdb::BackupInfo> backup_infos;
+  backup_engine->GetBackupInfo(&backup_infos);
+
+  if (!backup_infos.size()) {
```
```suggestion
  if (backup_infos.empty()) {
```

tchaikov reviewed (commented):
left a few comments.

frittentheke commented:
unstale

poelzi committed:
mon: add backup mechanism

Add the ability to create consistent backups of the monitor's RocksDB
database. Backups are exposed through the `backup` admin socket command
and performed by a MonitorBackupManager thread so the work runs off the
monitor's main thread.

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

poelzi committed:
mon: add backup restore mechanism

Add restoring a monitor RocksDB database from a backup, exposed through
the ceph-mon --restore-backup and --backup-version arguments.

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

poelzi committed:
mon: add auto backup option

Add a mon_auto_backup option which, when enabled, triggers a monitor
database backup on important changes to the auth and OSD state. The
monitors are notified of such changes via set_should_backup(), and the
pending request is serviced from the monitor tick.

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

poelzi committed:
mon: implement mon backup garbage collection

Add a garbage collection mechanism that prunes old monitor database
backups. backup_cleanup() keeps the last N backups plus one backup per
hourly and daily time bucket, and removes corrupted backups. It can be
triggered on demand via the "backup_cleanup" admin socket command or
periodically through mon_backup_cleanup_interval.

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

poelzi committed:
mon: add --list-backups argument

Add a --list-backups command line argument to ceph-mon which prints the
id, timestamp and size of every backup found in the given directory and
exits. The listing is implemented down the stack via
MonitorDBStore::list_backups() and RocksDBStore::list_backups().

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

poelzi committed:
doc: add documentation for mon backup and recovery

Document the monitor database backup feature: the backup and cleanup
mechanism and its tunables in the monitor configuration reference, and
the recovery procedure using --list-backups and --restore-backup in the
monitor troubleshooting guide.

Signed-off-by: Daniel Poelzleithner <poelzleithner@b1-systems.de>

tchaikov commented:
@poelzi Hi Daniel, would you be open to me adopting this pull request if you are no longer actively working on it?

I have already done the following on top of your work:

- Split the change into 7 self-contained, individually compilable commits mirroring your original structure
- Addressed all existing review comments
- Fixed typos in comments, config descriptions, and documentation
- Fixed perf-counter type/usage mismatches (backup_last_size/backup_last_files were declared as monotonic counters but used as gauges; backup_cleanup_started was the reverse)
- Fixed wrong doxygen `@param/@return` entries in MonitorDBStore
- Removed a stray dead variable and an unrelated file change

Your authorship will be preserved on all commits.

Please let me know if you have any objections or if you are still planning to continue the work yourself.

poelzi commented:
@tchaikov totally. I don't see myself finding the time to finish this. I hardly want to touch c/c++ anymore.

When testing, I found one bug back then. When starting a backup while the server was post pushing data, the dump was not binary exact. I think there was a extra stray byte at the end. I think it was a bug in the rocksdb backup code. If it still happens, I would ask opus to track it down - quite capable at those tasks

mheler commented:
@tchaikov I didn't see your update to this PR, would you like me to close mine at https://github.com/ceph/ceph/pull/68978 ?

tchaikov commented:
> @tchaikov I didn't see your update to this PR, would you like me to close mine at #68978 ?

@mheler I did update this pull request: I rebased its commits against the current main `HEAD` and implemented all the changes I proposed in my earlier comments. Could you take a look at the current state?

If you would like to take over the work, I am happy to hand it off. I would just ask that we avoid duplicating effort: I spent about two days reworking it, so it would be ideal if you could review the current shape first and then decide whether to continue from here or start fresh with your own PR.

tchaikov commented:
> @tchaikov totally. I don't see myself finding the time to finish this. I hardly want to touch c/c++ anymore.
> 
> When testing, I found one bug back then. When starting a backup while the server was post pushing data, the dump was not binary exact. I think there was a extra stray byte at the end. I think it was a bug in the rocksdb backup code. If it still happens, I would ask opus to track it down - quite capable at those tasks

@poelzi thanks! i manually tested `--list-backups` and `--restore-backup --backup-version 2` locally. they worked just fine. weird, we only bumped rocksdb submodule in e4d28e9fef6a5c0ed4c0675ea858e928ef2476f8 -- in july 2023 before you submitted this change.

mheler commented:
> > @tchaikov I didn't see your update to this PR, would you like me to close mine at #68978 ?
> 
> @mheler I did update this pull request: I rebased its commits against the current main `HEAD` and implemented all the changes I proposed in my earlier comments. Could you take a look at the current state?
> 
> If you would like to take over the work, I am happy to hand it off. I would just ask that we avoid duplicating effort: I spent about two days reworking it, so it would be ideal if you could review the current shape first and then decide whether to continue from here or start fresh with your own PR.

Thanks, @tchaikov ! I reviewed the current shape of https://github.com/ceph/ceph/pull/56772 and I appreciate the rebase/splitting work you put into it. After comparing the two, I’d prefer to continue with https://github.com/ceph/ceph/pull/68978 because it keeps the scope narrower (dropping mon_auto_backup) while adding restore safety, retention correctness, cleaner CLI/listing behavior, and backup scheduling/config cleanup that are not all present in https://github.com/ceph/ceph/pull/56772

tchaikov commented:
closing in favor of #68978
